### PR TITLE
[MM-14846] Update EditAt for FileIds and Attachment in Post + Ignore FileIds Updates

### PR DESCRIPTION
**Pull request by DSchalla**
#### Summary
This PR enforces that when the File IDs or Attachment property of a post is changed that the EditAt timestamp will be updated accordingly. Additionally, updates of the FileIds array will be ignored in the API handler.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-14846

### CODE REVIEW

File: app/post.go [hunk 1]
```diff
@@ -506,6 +507,11 @@ func (a *App) UpdatePost(post *model.Post, safeUpdate bool) (*model.Post, *model
 		newPost.Props = post.Props
 	}
 
+	// Avoid deep-equal checks if EditAt was already modified through message change
+	if newPost.EditAt == oldPost.EditAt && (!reflect.DeepEqual(oldPost.FileIds, newPost.FileIds) || !reflect.DeepEqual(oldPost.Attachments(), newPost.Attachments())) {
```
Comment (crspeller) [inline app/post.go]: The use of DeepEqual makes me a bit nervous here. Can we maybe use a custom comparator?

File: api4/post.go [hunk 2]
```diff
@@ -437,6 +437,9 @@ func updatePost(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// Updating the file_ids of a post is not a supported operation and will be ignored
+	post.FileIds = nil
```
Comment (crspeller) [inline api4/post.go]: Why is this necessary? It looks like the `UpdatePost(_, false)` should handle that.

File: api4/post.go [hunk 3]
```diff
@@ -479,6 +482,9 @@ func patchPost(c *Context, w http.ResponseWriter, r *http.Request) {
...
```
Comment (crspeller) [inline api4/post.go]: We also end up doing a `UpdatePost(_, false)` here.

File: api4/post.go (hunk 2 above)
Comment (DSchalla) [reply api4/post.go]: @crspeller When `safeUpdate == false`, `!false` --> `true`:
```
	if !safeUpdate || !false || true {
		newPost.IsPinned = post.IsPinned
		newPost.HasReactions = post.HasReactions
		newPost.FileIds = post.FileIds
		newPost.Props = post.Props
	}
```
When safeUpdate is true, it takes the values from the post being passed in the argument. When it's true, the original values will be maintained. (We seem to use that case only in test scenarios, it  is confusing to me as well.)

File: api4/post.go (hunk 3 above)
Comment (DSchalla) [reply api4/post.go]: See above.

**Comment by DSchalla** [comment]
@grundleborg You currently need to delete the entire post to do so, no other way is exposed in the UI from what I know.

File: api4/post.go (hunk 2 above)
Comment (crspeller) [reply api4/post.go]: Ugg. I hate these flags we sometimes have on functions that are super confusing.

Comment (DSchalla) [reply api4/post.go]: Yeah, it's quite confusing. I'll open a ticket to track it, because like you I was quite confused as well. I debugged it in the end to be sure.

File: app/post.go (hunk 1 above)
Comment (DSchalla) [reply app/post.go]: Yeah I think that's a good call, although it introduced ~448 LoC of code for both comparison operators + tests. I think since they are reusable and don't rely on reflection, that's a better way to go.

File: model/post.go [hunk 4]
```diff
@@ -403,6 +403,166 @@ func (o *Post) Attachments() []*SlackAttachment {
...
+			return false
+		}
+
+		// Compare length & slice values of fields
+
+		if len(attachments[i].Fields) != len(inputAttachments[i].Fields) {
+			return false
+		}
+
+		for j := range attachments[i].Fields {
```
Comment (crspeller) [inline model/post.go]: Maybe break these out into equality functions for their various structs? To avoid such a mega function.

Comment (DSchalla) [reply model/post.go]: Split up to each child function in the `Equals` function, makes it indeed nicer to read.
